### PR TITLE
[wip] replace whitelists with role auth

### PR DESCRIPTION
**Pull request by rainbreak**
Pull out the whitelists and replace them with [ds-roles](https://github.com/nexusdev/ds-roles), fixing #8.

This allows the owner fine grained access control via DSAuth, whilst providing a pre-configured default authority (via the factory) that follows the admin-issuer-holder pattern.

However, it doesn't entirely move function access into the `auth` modifier as `auth` isn't conditional on arguments: we still need e.g. `auth_recipient(address to)` to restrict holders from transferring to non-holders, which limits the configurability of the contract.

This isn't completely elegant - ideally we'd like to specify all access via the `auth` modifier. One solution to this is some hack by deserializing `msg.data` and using an internal lookup table. But that's not elegant either.

TODO:
- [x] tests
- [x] re-add modifier to prevent non holder transfer
- [x] update readme again
- [x] rebase
- [x] rethink ui
  - [x] display user roles
  - [x] display controls conditional on role
    - [x] register / unregister
    - [x] setFoo
    - [x] issue / cover
    - [x] transfer
  - [x] allow role updates by authority owner
- [x] resolve ds-roles .dapple/packages issue (maybe I can get 0.8 working...) [forked, for now]
- [x] remove ds-whitelist submodule
- [x] update js_modules.js
- [x] testnet deploy

### CODE REVIEW

**Comment by nmushegian** [comment]
You can make a custom modifier `needs_role(address who, uint8 what)` and do the same as in auth: `_authority.canCall(who, this, msg.sig)`

edit: I misread, looks like that is what you did. Right now I'm leaning towards there should be distinct modifiers for anything that is not sender-to-code relationship. In `canCall` the argument is called "caller", not "user"

File: src/simplecoin_test.sol [hunk 1]
```diff
@@ -61,13 +50,12 @@ contract SimplecoinTest is Test {
...
-        var coin = factory.create(feedbase, "some rules", issuers, holders);
-        assertEq(this, coin.owner());
+        var coin = factory.create(feedbase, "some rules");
         // TODO: check authority setup
     }
 
-    function testCreatorIsOwner() {
-        assertEq(this, coin.owner());
+    function testCreatorHasAuthority() {
+        assertEq(this, coin._authority());
```
Comment (nmushegian) [inline src/simplecoin_test.sol:58]: want to also check the mode. see discussion in issues

Comment (rainbreak) [reply src/simplecoin_test.sol:58]: Yes. The tests I haven't pushed are much more thorough.

**Comment by rainbreak** [comment]
A useful addition to ds-roles would be `hasRole(address who, uint8 role) returns (bool)` (see https://github.com/makerdao/simplecoin/pull/26/commits/19b8708239dadea0271f24b49c1c5c8956f9d35c#diff-23035d25bf7b473d1d2065f3b2c64961R80)

Stated more clearly, my issue with modifiers here is that they put awareness of the intended authentication scheme into the contract. If transfer isn't allowed to non-holders, then it needs to 1) be aware of that with a modifier (which is fine) and 2) know that the authority is a `DSRoleAuth` so that it can check.

Again, this is all ok it just limits the authentication configuration a bit. The simplecoin with whitelists knows all about the whitelists so it isn't really a problem for this simplecoin to know all about the role-auth.

**Comment by rainbreak** [comment]
Ok, non-holders can again be blocked from receiving coin. The authority just needs to implement `canReceive(address who, uint amount) returns bool` (as well as `canCall`).

**Comment by dbrock** [comment]
I'm sorry, I pushed some stuff that conflicts with this (#25). I'm pretty sure we should just go with your version for the conflicting parts.

**Comment by nmushegian** [comment]
@rainbeam I don't think adding methods to the authority is right. If it doesn't fit the canCall mold then canReceive interpreting should be local and hide how it "abuses" canCall

**Comment by rainbreak** [comment]
@nmushegian: I was aiming to minimise auth constraints in the simplecoin by leaving it up to the authority how to implement `canReceive`. I see now that I could abuse canCall for this purpose - by asserting that the transfer recipient can also call transfer.

Is it your view that authorities should only expose `canCall` to the contracts that they authorise?

**Comment by nmushegian** [comment]
> Is it your view that authorities should only expose canCall to the contracts that they authorise?

Yes, `canReceive` is far too specific - if we go this route then this contract should inherit from some subclass of `DSAuth`

**Comment by nmushegian** [comment]
I do like putting `hasRole` on the role auth though

**Comment by rainbreak** [comment]
ok, removed canReceive.

It might be that preventing non-holders from receiving simplecoin is too restrictive. It's not like they can do anything with it anyway. I suppose they could sell their private keys.

**Comment by rainbreak** [comment]
Rebased and updated readme. Now need to think about how to reflect the role auth in the ui.

**Comment by rainbreak** [comment]
Exposed a fair bit of api:



So now there's technically four roles - "admin", "issuer", "holder" as before and "owner" which is actually the authority of the coin authority.

**Comment by rainbreak** [comment]
Ok that'll do for now. I'm pretty sure the ui is broken with metamask somehow, but can fix that elsewhere. This at least showcases the role structure.

**Pull request merged**